### PR TITLE
Adding detail for openid connect configuration section

### PR DESCRIPTION
**Pull request by mlbiam**
Integrated sections of https://github.com/TremoloSecurity/wiki/blob/master/kubernetes.md related to oidc integration that others have found useful removing product specific commentary and adding OpenUnison as a possible identity provider

cc @kubernetes/sig-auth

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/2089)

### CODE REVIEW

**Comment by mlbiam** [comment]
I signed it!

File: docs/admin/authentication.md [hunk 1]
```diff
@@ -222,44 +222,103 @@ from the OAuth2 [token response](https://openid.net/specs/openid-connect-core-1_
...
+
+1.  Support [OpenID connect discovery](https://openid.net/specs/openid-connect-discovery-1_0.html); not all do.
+2.  Run in TLS with non-obsolete ciphers
+3.  Have a CA signed certificate (even if the CA is not a commercial CA or is self signed)
 
-The provider needs to support [OpenID connect discovery](https://openid.net/specs/openid-connect-discovery-1_0.html); not all do.
+A note about requirement #3 above, requiring a CA signed certificate.  If you deploy your own identity provider (as apposed to one of the cloud providers like Goolge or Microsoft) you MUST have your identity provider's web server certificate signed by a CA.  This is an issue with GoLang's TLS client implementation not being able to verify a self-signed certificate.  If you don't have a CA handy, you can use this script from the CoreOS team to create a simple CA and a signed certificate and key pair - https://github.com/coreos/dex/blob/1ee5920c54f5926d6468d2607c728b71cfe98092/examples/k8s/gencert.sh or this script based on it that will generate sha256 certs with a longer life and larger keysize https://raw.githubusercontent.com/TremoloSecurity/openunison-qs-kubernetes/master/makecerts.sh.
```
Comment (liggitt) [inline docs/admin/authentication.md]: > This is an issue with GoLang's TLS client implementation not being able to verify a self-signed certificate.

that's not strictly true... a self-signed cert works as long as the cert is allowed to be used as a CA (CA:TRUE, etc)

File: docs/admin/authentication.md [hunk 2]
```diff
@@ -222,44 +222,103 @@ from the OAuth2 [token response](https://openid.net/specs/openid-connect-core-1_
...
+##### Option 1 - OIDC Authenticator
+
+The first option is to use the oidc authenticator.  This authenticator takes your `id_token`, `refresh_token` and your OIDC `client_secret` and will refresh your token automatically.  Once you have authenticated to your identity provider:
+
+```
+kubectl config set-credentials USER_NAME --auth-provider=oidc
+kubectl config set-credentials USER_NAME --auth-provider-arg=idp-issuer-url=( issuer url )
+kubectl config set-credentials USER_NAME --auth-provider-arg=client-id=( your client id )
+kubectl config set-credentials USER_NAME --auth-provider-arg=client-secret=( your client secret )
+kubectl config set-credentials USER_NAME --auth-provider-arg=refresh-token=( your refresh token )
```
Comment (liggitt) [inline docs/admin/authentication.md]: can't all of this be done in a single command?

File: docs/admin/authentication.md [hunk 3]
```diff
@@ -222,44 +222,103 @@ from the OAuth2 [token response](https://openid.net/specs/openid-connect-core-1_
...
+```
+Alternatively, you can add these configuration options manually to your `~/.kube/config`:
```
Comment (liggitt) [inline docs/admin/authentication.md]: if we include this, I would only do so as an example of the resulting file generated from the `set-credentials` command

File: docs/admin/authentication.md (hunk 1 above)
Comment (mlbiam) [inline docs/admin/authentication.md]: ok, i'll make an update

File: docs/admin/authentication.md (hunk 2 above)
Comment (mlbiam) [inline docs/admin/authentication.md]: I'll try it out

File: docs/admin/authentication.md (hunk 3 above)
Comment (mlbiam) [inline docs/admin/authentication.md]: @liggitt so you want me to show running the kubectl command that creates the config and the resulting config?

Comment (liggitt) [inline docs/admin/authentication.md]: yes, if we show the resulting kubeconfig contents at all. I suppose it's interesting to show, but I wouldn't expect manual manipulation to be the recommended way to set it.

**Comment by mlbiam** [comment]
@liggitt did my updates take care of your concerns?

File: docs/admin/authentication.md [hunk 4]
```diff
@@ -222,44 +222,122 @@ from the OAuth2 [token response](https://openid.net/specs/openid-connect-core-1_
...
+#### Configuring the API Server
+
+To enable the plugin, configure the following flags on the API Server:
+
+| Parameter | Description | Example | Required |
+| --------- | ----------- | ------- | ------- |
+| --oidc-issuer-url | URL of the provider which allows the API server to discover public signing keys. Only URLs which use the `https://` scheme are accepted.  This is typically the provider's discovery URL without a path, for example "https://accounts.google.com" or "https://login.salesforce.com".  This URL should point to the level below .well-known/openid-configuration | If the discovery URL is https://accounts.google.com/.well-known/openid-configuration the value should be https://accounts.google.com | Yes |
+| --oidc-client-id |  A client id that all tokens must be issued for. | kubernetes | Yes |
+| --oidc-username-claim | JWT claim to use as the user name. By default `sub`, which is expected to be a unique identifier of the end user. Admins can choose other claims, such as `email`, depending on their provider. | sub | No |
+| --oidc-groups-claim | JWT claim to use as the user's group. If the claim is present it must be an array of strings. | user_roles | No |
```
Comment (liggitt) [inline docs/admin/authentication.md]: `groups` might be a more typical example

**Comment by mlbiam** [comment]
@liggitt ok, changed user_roles --> groups

File: docs/admin/authentication.md [hunk 5]
```diff
@@ -222,44 +222,122 @@ from the OAuth2 [token response](https://openid.net/specs/openid-connect-core-1_
...
+1.  Support [OpenID connect discovery](https://openid.net/specs/openid-connect-discovery-1_0.html); not all do.
+2.  Run in TLS with non-obsolete ciphers
+3.  Have a CA signed certificate (even if the CA is not a commercial CA or is self signed)
+
+A note about requirement #3 above, requiring a CA signed certificate.  If you deploy your own identity provider (as apposed to one of the cloud providers like Goolge or Microsoft) you MUST have your identity provider's web server certificate signed by a certificate with the `CA` flag set to `TRUE`, even if it is self signed.  This is due to GoLang's TLS client implementation being very strict to the standards around certificate validation.  If you don't have a CA handy, you can use this script from the CoreOS team to create a simple CA and a signed certificate and key pair - https://github.com/coreos/dex/blob/1ee5920c54f5926d6468d2607c728b71cfe98092/examples/k8s/gencert.sh or this script based on it that will generate sha256 certs with a longer life and larger keysize https://raw.githubusercontent.com/TremoloSecurity/openunison-qs-kubernetes/master/makecerts.sh.
```
Comment (liggitt) [inline docs/admin/authentication.md]: sorry, just noticed typos "apposed" and "Goolge"

**Comment by mlbiam** [comment]
@liggitt just realized the edit window on github doesn't have spell check enabled.  Fixed those and a couple of minor case errors too

File: docs/admin/authentication.md [hunk 6]
```diff
@@ -222,44 +222,122 @@ from the OAuth2 [token response](https://openid.net/specs/openid-connect-core-1_
...
-* `--oidc-username-claim` JWT claim to use as the user name. By default `sub`,
-which is expected to be a unique identifier of the end user. Admins can choose
-other claims, such as `email`, depending on their provider.
-* `--oidc-groups-claim` JWT claim to use as the user's group. If the claim is present
-it must be an array of strings.
-
 Kubernetes does not provide an OpenID Connect Identity Provider.
 You can use an existing public OpenID Connect Identity Provider (such as Google, or [others](http://connect2id.com/products/nimbus-oauth-openid-connect-sdk/openid-connect-providers)).
-Or, you can run your own Identity Provider, such as CoreOS [dex](https://github.com/coreos/dex), [Keycloak](https://github.com/keycloak/keycloak) or CloudFoundry [UAA](https://github.com/cloudfoundry/uaa).
+Or, you can run your own Identity Provider, such as CoreOS [dex](https://github.com/coreos/dex), [Keycloak](https://github.com/keycloak/keycloak), CloudFoundry [UAA](https://github.com/cloudfoundry/uaa), or Tremolo Secutity's [OpenUnison](https://github.com/tremolosecurity/openunison).
```
Comment (liggitt) [inline docs/admin/authentication.md]: typo: Secutity

File: docs/admin/authentication.md [hunk 7]
```diff
@@ -222,44 +222,122 @@ from the OAuth2 [token response](https://openid.net/specs/openid-connect-core-1_
...
+![Kubernetes OpenID Connect Flow](/images/docs/admin/k8s_oidc_login.svg)
 
-* `--oidc-issuer-url` URL of the provider which allows the API server to discover
-public signing keys. Only URLs which use the `https://` scheme are accepted.  This is typically
-the provider's URL without a path, for example "https://accounts.google.com" or "https://login.salesforce.com".
+1.  Login to your identity provider
+2.  Your identity provider will provide you with an `access_token`, `id_token` and a `refresh_token`
+3.  When using `kubectl`, use your `id_token` with the `--token` flag or add it directly to your `kubeconfig`
+4.  `kubectl` sends your `id_token` in a header called Authorization to the API Server
+5.  The api server will make sure the JWT signature is valid by checking against the certificate named in the configuration
```
Comment (liggitt) [inline docs/admin/authentication.md]: API server

File: docs/admin/authentication.md [hunk 8]
```diff
@@ -222,44 +222,122 @@ from the OAuth2 [token response](https://openid.net/specs/openid-connect-core-1_
...
-To enable the plugin, pass the following required flags:
+![Kubernetes OpenID Connect Flow](/images/docs/admin/k8s_oidc_login.svg)
 
-* `--oidc-issuer-url` URL of the provider which allows the API server to discover
-public signing keys. Only URLs which use the `https://` scheme are accepted.  This is typically
-the provider's URL without a path, for example "https://accounts.google.com" or "https://login.salesforce.com".
+1.  Login to your identity provider
+2.  Your identity provider will provide you with an `access_token`, `id_token` and a `refresh_token`
+3.  When using `kubectl`, use your `id_token` with the `--token` flag or add it directly to your `kubeconfig`
+4.  `kubectl` sends your `id_token` in a header called Authorization to the API Server
```
Comment (liggitt) [inline docs/admin/authentication.md]: API server

**Comment by liggitt** [comment]
sorry for the nits... go ahead and squash then LGTM for tech review

**Comment by mlbiam** [comment]
@liggitt fixed those and another typo i found

**Comment by liggitt** [comment]
tech review LGTM, go ahead and squash

**Comment by mlbiam** [comment]
@liggitt sorry for the dumb q, but do I need to squash or is that done on the merge?

**Comment by liggitt** [comment]
Not sure about this repo, but the main one doesn't squash for you

**Comment by mlbiam** [comment]
@liggitt ok so I feel like I'm missing something really simple here.  I can't get git to squash for the life of me.  I've tried multiple variations of:

```
$ git rebase -i origin/master
```

then putting in 

```
p 14b7375 initial checkin
s 922b0b2 integrated from ts wiki
s 405085f updated to point to graphic in git
s 20cf69b updated for correct styles
s 307b384 removed prompts
s 08c9437 Merge branch 'master' into master
s c177ff7 Merge branch 'master' into master
s 85dde65 updated per @liggitt's comments
s 2f23edd added specific examples per @liggitt's request
s 6d7f801 Merge branch 'master' into master
s 02f7867 Merge branch 'master' into master
s 8406921 Merge branch 'master' into master
s db85be7 Merge branch 'master' into master
s 5b82ebf Merge branch 'master' into master
s 49f1015 Merge branch 'master' into master
s b9576ff changed user_roles --> groups
s 4df38b8 minor spelling and case corrections
s bbe307e Merge branch 'master' into master
s ed3ddb2 Merge branch 'master' into master
s 3812f4a Merge branch 'master' into master
s d9fb8ea fixed last few typos
s a1e7cd3 found another typo
s 95cb7b8 Merge branch 'master' into master
s 219bb25 Merge branch 'master' into master
```

and I get the following:

```
The previous cherry-pick is now empty, possibly due to conflict resolution.
If you wish to commit it anyway, use:

    git commit --allow-empty

Otherwise, please use 'git reset'
interactive rebase in progress; onto 219bb25
Last command done (1 command done):
   p 14b7375 initial checkin
Next commands to do (23 remaining commands):
   s 922b0b2 integrated from ts wiki
   s 405085f updated to point to graphic in git
You are currently rebasing branch 'master' on '219bb25'.

nothing to commit, working tree clean
Could not apply 14b73758922dfcdb7ff7c0e19ae41207995474d8... initial checkin
```

I've tried forward, backwords, no merge...i feel like i'm missing something very obvious.  Any help would be greatly appreciated.

Thanks

**Pull request closed without merging**